tv_apps/smart-system/js/focus_manager.js

```diff
@@ -0,0 +1,127 @@
+/* global TrustedUIManager, sleepMenu, AppWindowManager, ScreenManager */
+
+(function(exports) {
+  'use strict';
+
+  // The following keys are handled by state machine. If we prevent the
+  // following keys, we may break the state machine. Before a total solution,
+  // we do not use the following keys as a fallback trigger.
+  var BLACK_LIST_KEYS = [
+    'power', // power key will be handled by screen manager.
+    'home', // home key will be handled by homescreen manager which sets focus
+            // to correct place.
+    'mozhomescreen',
+    'volumeup',
+    'volumedown'
+  ];
+
+  function FocusManager() {
+  }
+
+  var proto = FocusManager.prototype;
+
+  proto.start = function ff_start() {
+    this.startFallbackCatcher();
+  };
+
+  proto.stop = function ff_stop() {
+    this.stopFallbackCatcher();
+  };
+
+  proto.focus = function ff_focus() {
+    var systemUIList = [
+      TrustedUIManager,
+      sleepMenu
+    ];
+    // list all visible system UI
+    var visible = systemUIList.filter(function(item) {
+      return item && item.isVisible();
+    });
+    if (visible.length > 0) {
+      // sort system UIs
+      visible.sort(function(item1, item2) {
+        return item2.getOrder() - item1.getOrder();
+      });
+      // focus top-most system UI
+      visible[0].focus();
+    } else {
+      // no system UI, we set focus back to top-most AppWindow
+      AppWindowManager.getActiveApp().focus();
+    }
+
+  };
+
+  proto.startFallbackCatcher = function ff_startFallbackCatcher() {
+    window.addEventListener('keydown', this, true);
+    window.addEventListener('keyup', this, true);
+    window.addEventListener('mozbrowserbeforekeydown', this, true);
+    window.addEventListener('mozbrowserbeforekeyup', this, true);
+  };
+
+  proto.stopFallbackCatcher = function ff_stopFallbackCatcher() {
+    window.removeEventListener('keydown', this, true);
+    window.removeEventListener('keyup', this, true);
+    window.removeEventListener('mozbrowserbeforekeydown', this, true);
+    window.removeEventListener('mozbrowserbeforekeyup', this, true);
+  };
+
+  proto.isActiveElementVisible = function ff_isActiveElementVisible() {
+    var active = document.activeElement;
+    var style = active ? window.getComputedStyle(active) : null;
+    var visible = true;
+    // check its ancestor is all visible.
+    while(active && active.tagName !== 'HTML') {
+      style = window.getComputedStyle(active);
+      if (style.display === 'none' || style.visibility === 'hidden') {
```

Comment:
We may need to clarify the visibility of AppIframe (Although it works now since their parent sets visibility to hidden)